### PR TITLE
Feature 2476 tc pairs diag

### LINKED ISSUE #2476: Enhance TC-Pairs to include storm diagnostics in consensus track output
## Describe the Enhancement ##
When the `-diag` command line option is provided, the TC-Pairs tool writes TCDIAG output lines after the TCMPR lines to which they correspond. However these diagnostics are NOT currently included in the computation of consensus tracks.

This issue is to enhance the computation of consensus tracks to include the average of the model diagnostics of the members. There are obvious questions about whether or not each diagnostic must be present and valid in the members for it to be included in the consensus output. Does this need to be configurable?

### Time Estimate ###
2 days?

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
None needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2770043

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
None.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by sethlinden**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>

The changes add consensus diagnostic values to the consensus output file

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

I used the new tc_pairs unit test: tc_pairs_DIAGNOSTICS_CONSENSUS to formulate my own command line:

```
./tc_pairs -adeck /d1/projects/MET/MET_test_data/unit_test/tc_data/adeck/aal092022.dat -bdeck /d1/projects/MET/MET_test_data/unit_test/tc_data/bdeck/bal092022.dat -diag CIRA_DIAG_RT /d1/projects/MET/MET_test_data/unit_test/tc_data/diag/cira_diag_rt/2022/GFSI/sal092022_avni_doper_20220926*_diag.dat -diag CIRA_DIAG_RT /d1/projects/MET/MET_test_data/unit_test/tc_data/diag/cira_diag_rt/2022/EMXI/sal092022_emxi_doper_20220926*_diag.dat -config /d1/personal/linden/git/feature_2476_tc_pairs_diag/MET/internal/test_unit/config/TCPairsConfig_DIAGNOSTICS_CONSENSUS -out al092022_20220926_DIAGNOSTICS_CONSENSUS
```

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Run the unit test unit_tc_pairs.xml and look specifically at the output for the tc_pairs_DIAGNOSTICS_CONSENSUS unit test

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [x] Do these changes include sufficient testing updates? **[Yes]**

Yes a new unit test was added to unit_tc_pairs.xml (tc_pairs_DIAGNOSTICS_CONSENSUS)

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

Yes, this change added consensus (average) diagnostic values to the output file

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [ ] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: internal/test_unit/xml/unit_tc_pairs.xml [hunk 1]
```diff
@@ -206,4 +206,27 @@
     </output>
   </test>
 
+  <test name="tc_pairs_DIAGNOSTICS_CONSENSUS">
```
Comment (JohnHalleyGotway) [inline internal/test_unit/xml/unit_tc_pairs.xml:209]: Note that this remains not a very good test. Recommend replacing this with a more realistic test of the functionality in future work.

File: src/libcode/vx_tc_util/track_info.cc [hunk 2]
```diff
@@ -562,7 +562,7 @@ bool TrackInfo::add_diag_data(DiagFile &diag_file, const StringArray &req_diag_n
       // Store the diagnostic name
       DiagName.add(diag_file.diag_name()[i_name]);
       NumArray diag_val = diag_file.diag_val(diag_file.diag_name()[i_name]);
-
+
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion

```

File: src/libcode/vx_tc_util/track_info.cc [hunk 3]
```diff
@@ -965,11 +975,11 @@ int TrackInfoArray::add_diag_data(DiagFile &diag_file, const StringArray &diag_n
 TrackInfo consensus(const TrackInfoArray &tracks,
                     const ConcatString &model, int req,
                     const StringArray &req_list) {
-   int i, j, i_pnt;
+   int i, j, k, i_pnt;
    bool skip;
    TrackInfo tavg;
    NumArray lead_list;
-
+
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion

```

File: src/libcode/vx_tc_util/track_info.cc [hunk 4]
```diff
@@ -993,7 +1003,7 @@ TrackInfo consensus(const TrackInfoArray &tracks,
    tavg.set_technique(model.c_str());
    tavg.set_init(tracks[0].init());
    tavg.set_storm_id();
-
+
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion

```

File: src/libcode/vx_tc_util/track_info.cc [hunk 5]
```diff
@@ -1016,6 +1026,13 @@ TrackInfo consensus(const TrackInfoArray &tracks,
               << tracks[i].technique_number() << ").\n\n";
       }
 
+      // Get unique diag names
+      // Store diag_names in tavg object
+      for(j=0; j<tracks[i].diag_name().n(); j++) {
+         mlog << Debug(3) << "TrackInfo::consensus(). Found unique diag_name[" << j << "] = " << tracks[i].diag_name()[j] << "\n";
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion
```
I think I'd just remove this misleading log message. The diag_name will be logged for all input tracks and stating them as being unique is misleading.

File: src/libcode/vx_tc_util/track_info.cc [hunk 6]
```diff
@@ -1056,11 +1073,18 @@ TrackInfo consensus(const TrackInfoArray &tracks,
             }
             continue;
          }
-
+
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion

```

File: src/libcode/vx_tc_util/track_info.cc [hunk 7]
```diff
@@ -1138,6 +1162,52 @@ TrackInfo consensus(const TrackInfoArray &tracks,
       // Compute consensus CycloneLevel
       if(!is_bad_data(pavg.v_max())) pavg.set_level(wind_speed_to_cyclonelevel(pavg.v_max()));
 
+      // Loop over the diag name and the input track points and get consensus diag value
+      // (mean value across all members that have the diag value)
+      for (j=0; j<tavg.diag_name().n(); j++){
+         mlog << Debug(3) << "TrackInfo::consensus(). Working on calculating mean values for tavg.diag_name()[" << j << "] = " << tavg.diag_name()[j] << "\n";
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion
         mlog << Debug(4) << "Computing consensus \"" << tavg.diag_name[j] << "\" 
 diagnostic value.\n";
```
Making log message more concise and increasing to verbosity level 4.

File: src/libcode/vx_tc_util/track_info.cc [hunk 8]
```diff
@@ -1138,6 +1162,52 @@ TrackInfo consensus(const TrackInfoArray &tracks,
...
+               if(req_list.has(tracks.Track[k].technique())) {
+                  skip = true;
+                  break;
+               }
+               continue;
+            }
+
+            // Add non-missing diag values to local NumArray 
+            if(!is_bad_data( tracks.Track[k][i_pnt].get_diag_val(tavg.diag_name(), tavg.diag_name()[j]) )) {
+               mlog << Debug(3) <<  "  Adding values for k: " << k << ", i_pnt: " << i_pnt << " tracks.Track[k][i_pnt].get_diag_val(tavg.diag_name(), tavg.diag_name()[j]) = " << tracks.Track[k][i_pnt].get_diag_val(tavg.diag_name(), tavg.diag_name()[j]) << "\n";
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion
```
This looks like debugging code useful for development and overkill to clutter the log file with in general.

File: src/libcode/vx_tc_util/track_info.cc [hunk 9]
```diff
@@ -1138,6 +1162,52 @@ TrackInfo consensus(const TrackInfoArray &tracks,
...
+               diag_vals.add(tracks.Track[k][i_pnt].get_diag_val(tavg.diag_name(), tavg.diag_name()[j]));
+            }
+            
+         } // end loop over ensemble of tracks (for given lead-time: i_pnt)
+         
+         cons_diag_val = diag_vals.mean();
+         mlog << Debug(3) << "  cons_diag_val = " << cons_diag_val << "\n";
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion
```
This looks like debugging code useful for development and overkill to clutter the log file with in general.

File: src/libcode/vx_tc_util/track_info.cc [hunk 10]
```diff
@@ -1138,6 +1162,52 @@ TrackInfo consensus(const TrackInfoArray &tracks,
...
+         
+         // Add the cons_diag_val to the pavg DiagVal NumArray (using add_diag_value)
+         if(!is_bad_data(cons_diag_val)) {
+            mlog << Debug(3) << "  Adding cons_diag_val to pavg DiagVal array" << "\n";
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]:
```suggestion
```
This looks like debugging code useful for development and overkill to clutter the log file with in general.

File: src/libcode/vx_tc_util/track_info.h [hunk 11]
```diff
@@ -169,7 +169,9 @@ class TrackInfo {
       void add_watch_warn(const ConcatString &, WatchWarnType, unixtime);
       bool add_diag_data(DiagFile &, const StringArray &);
       void add_diag_value(int, double);
-
+      //void add_uniq_diag_name(const string diag_name);
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.h]:
```suggestion
```
Deleting stale commented out code.

File: src/libcode/vx_tc_util/track_info.cc [hunk 12]
```diff
@@ -1138,6 +1162,52 @@ TrackInfo consensus(const TrackInfoArray &tracks,
...
+               diag_vals.add(tracks.Track[k][i_pnt].get_diag_val(tavg.diag_name(), tavg.diag_name()[j]));
+            }
+            
+         } // end loop over ensemble of tracks (for given lead-time: i_pnt)
+         
+         cons_diag_val = diag_vals.mean();
+         mlog << Debug(3) << "  cons_diag_val = " << cons_diag_val << "\n";
+         
+         // Add the cons_diag_val to the pavg DiagVal NumArray (using add_diag_value)
+         if(!is_bad_data(cons_diag_val)) {
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_tc_util/track_info.cc]: Don't we actually NEED to store the computed diagnostic value regardless of whether or not it contains bad data? For example, let's say all the input tracks report a diag value with bad data. If we fail to write bad data to the output, we'll end up with a mismatch of diagnostic names and values.

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
Seth, thanks for the PR. I proposed several cleanup changes, removing whitespace, commented-out code, and overly detail log messages that are useful for development but overkill for users.

The only substantive change is the handling of missing data. I think we should store a computed diagnostic value for each diagnostic name, regardless of whether it contains good or bad data. If we don't store bad data, we'll end up with a mismatch of diagnostic names and values.

File: src/libcode/vx_tc_util/track_info.cc (hunk 12 above)
Comment (sethlinden) [reply src/libcode/vx_tc_util/track_info.cc]: Hi John, I wasn't sure if we wanted to save the missing data. But based on what you are saying it makes sense.

**Comment by sethlinden** [comment]
@JohnHalleyGotway thanks for the thorough review and proposing the code changes. All your changes look reasonable to me including removing some of the irrelevant debug / log messages. What's the best way to proceed? Do I just look at the suggested changes, here (via GitHub online) and then accept the changes (to get them into the feature branch)?

**Comment by JohnHalleyGotway** [comment]
Seth, yep, you could use the "commit suggestion" button next to each change. Although that may trigger many runs of the GHA testing workflow. And I didn't make a code suggestion for stashing the diag value even if its bad data. So you'll need to make that change in your local branch and push it.

Because of that, I'd recommend just making all these changes locally, doing a single commit, and then pushing up that commit. That'll prevents lots of unneeded GHA runs.

**Comment by sethlinden** [comment]
@JohnHalleyGotway I basically did a combination of both. I used the "commit suggestion" for most of the changes above (blank spaces, debug changes)....on a side note it didn't initially like one of your suggested debug/log changes. I fixed it in my branch (from seneca).

Then I also update track_info.cc to save missing consensus debug values as well. I pushed the changes to GitHub.

I then merged latest from develop. And did a new "make clean install test".

Then I re-ran my own tc_pairs test. Everything looks good. I think this is close to being set now. Can you confirm?

**Comment by sethlinden** [comment]
@JohnHalleyGotway the workflow testing is still failing so I need to look into that next. Once we have confirmed the code changes. I assume the changes you made to the config file and documentation made it into this branch as well (since I didn't see a  commit suggestion for those changes).

**Comment by JohnHalleyGotway** [comment]
@sethlinden sorry, I missed the notification for this the other day. I realize that I never re-created the input test data tarfile to contain the contrived test data needed for this new tc_pairs unit test. I'd been hoping we'd replace the test, but haven't gotten there yet. So I'm recreating that tar file on mohawk now and will rerun the tests for this PR when that's done.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes. Thanks for bearing with me @sethlinden. After updating the GHA input dataset, I note that the unit tests all ran without error.

The new TC-Pairs output file is flagged as a difference and `develop-ref` will need be updated to get past that change.

I'll go ahead and push the button to merge these changes.

**Pull request merged**